// @babylonjs/core/FlowGraph/Blocks/Execution/Animation/flowGraphPlayAnimationBlock.js
import { FlowGraphAsyncExecutionBlock } from "../../../flowGraphAsyncExecutionBlock.js";
import { RichTypeAny, RichTypeNumber, RichTypeBoolean } from "../../../flowGraphRichTypes.js";
import { RegisterClass } from "../../../../Misc/typeStore.js";
/**
 * @experimental
 * A block that plays an animation on an animatable object.
 */
export class FlowGraphPlayAnimationBlock extends FlowGraphAsyncExecutionBlock {
    constructor(config) {
        super(config);
        this.target = this._registerDataInput("target", RichTypeAny);
        this.animation = this._registerDataInput("animation", RichTypeAny);
        this.speed = this._registerDataInput("speed", RichTypeNumber);
        this.loop = this._registerDataInput("loop", RichTypeBoolean);
        this.from = this._registerDataInput("from", RichTypeNumber);
        this.to = this._registerDataInput("to", RichTypeNumber);
        this.onAnimationEnd = this._registerSignalOutput("onAnimationEnd");
        this.runningAnimatable = this._registerDataOutput("runningAnimatable", RichTypeAny);
    }
    /**
     * @internal
     * @param context
     */
    _preparePendingTasks(context) {
        var _a;
        const targetValue = this.target.getValue(context);
        const animationValue = this.animation.getValue(context);
        if (!targetValue || !animationValue) {
            throw new Error("Cannot play animation without target or animation");
        }
        const contextAnimatables = (_a = context._getExecutionVariable(this, "runningAnimatables")) !== null && _a !== void 0 ? _a : [];
        // was an animation started on this target already and was just paused? if so, we can unpause it.
        const existingAnimatable = this.runningAnimatable.getValue(context);
        if (existingAnimatable && existingAnimatable.paused) {
            existingAnimatable.restart();
        }
        else {
            const scene = context.configuration.scene;
            const animatable = scene.beginDirectAnimation(targetValue, [animationValue], this.from.getValue(context), this.to.getValue(context), this.loop.getValue(context), this.speed.getValue(context), () => this._onAnimationEnd(animatable, context));
            this.runningAnimatable.setValue(animatable, context);
            contextAnimatables.push(animatable);
        }
        context._setExecutionVariable(this, "runningAnimatables", contextAnimatables);
    }
    _execute(context) {
        this._startPendingTasks(context);
        this.onDone._activateSignal(context);
    }
    _onAnimationEnd(animatable, context) {
        var _a;
        const contextAnims = (_a = context._getExecutionVariable(this, "runningAnimatables")) !== null && _a !== void 0 ? _a : [];
        const index = contextAnims.indexOf(animatable);
        if (index !== -1) {
            contextAnims.splice(index, 1);
        }
        context._removePendingBlock(this);
        this.onAnimationEnd._activateSignal(context);
    }
    /**
     * @internal
     * Stop any currently running animations.
     */
    _cancelPendingTasks(context) {
        var _a;
        const contextAnims = (_a = context._getExecutionVariable(this, "runningAnimatables")) !== null && _a !== void 0 ? _a : [];
        for (const anim of contextAnims) {
            anim.stop();
        }
        context._deleteExecutionVariable(this, "runningAnimatables");
    }
    getClassName() {
        return "FGPlayAnimationBlock";
    }
}
RegisterClass("FGPlayAnimationBlock", FlowGraphPlayAnimationBlock);
