// @babylonjs/core/FlowGraph/serialization.js
import { Quaternion, Vector2, Vector3, Vector4 } from "../Maths/math.vector.js";
function isMeshClassName(className) {
    return (className === "Mesh" ||
        className === "AbstractMesh" ||
        className === "GroundMesh" ||
        className === "InstanceMesh" ||
        className === "LinesMesh" ||
        className === "GoldbergMesh" ||
        className === "GreasedLineMesh" ||
        className === "TrailMesh");
}
function isVectorClassName(className) {
    return className === "Vector2" || className === "Vector3" || className === "Vector4" || className === "Quaternion";
}
function parseVector(className, value) {
    if (className === "Vector2") {
        return Vector2.FromArray(value);
    }
    else if (className === "Vector3") {
        return Vector3.FromArray(value);
    }
    else if (className === "Vector4") {
        return Vector4.FromArray(value);
    }
    else if (className === "Quaternion") {
        return Quaternion.FromArray(value);
    }
    else {
        throw new Error(`Unknown vector class name ${className}`);
    }
}
/**
 * The default function that serializes values in a context object to a serialization object
 * @param key the key where the value should be stored in the serialization object
 * @param value the value to store
 * @param serializationObject the object where the value will be stored
 */
export function defaultValueSerializationFunction(key, value, serializationObject) {
    var _a, _b;
    const className = (_b = (_a = value === null || value === void 0 ? void 0 : value.getClassName) === null || _a === void 0 ? void 0 : _a.call(value)) !== null && _b !== void 0 ? _b : "";
    if (isMeshClassName(className)) {
        serializationObject[key] = {
            name: value.name,
            className,
        };
    }
    else if (isVectorClassName(className)) {
        serializationObject[key] = {
            value: value.asArray(),
            className,
        };
    }
    else {
        serializationObject[key] = value;
    }
}
/**
 * The default function that parses values stored in a serialization object
 * @param key the key to the value that will be parsed
 * @param serializationObject the object that will be parsed
 * @param scene
 * @returns
 */
export function defaultValueParseFunction(key, serializationObject, scene) {
    const value = serializationObject[key];
    let finalValue;
    const className = value === null || value === void 0 ? void 0 : value.className;
    if (isMeshClassName(className)) {
        finalValue = scene.getMeshByName(value.name);
    }
    else if (isVectorClassName(className)) {
        finalValue = parseVector(className, value.value);
    }
    else {
        finalValue = value;
    }
    return finalValue;
}
